### PR TITLE
Carets- Julia Meier- Hotel

### PR DESCRIPTION
**Pull request by julmeier**
# Hotel
Congratulations! You're submitting your assignment!

## Comprehension Questions
| Question | Answer    |
| :------------- | :------------- |
| Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? | In creating the initial design, I debated as to which class should handle making reservations. I ultimately chose the Hotel class to check, store and modify reservations, because I wanted it to mimic real life in that, "a hotel makes reservations/ has rooms". The reservation, block and room classes did not take on any of this functionality. I could use some practice on how to handle exceptions.  |
| Describe a concept that you gained more clarity on as you worked on this assignment. | This assignment really helped me understand how and when to make objects, and how to call them from other classes and make modifications. 
I wish I could have gotten to the optional assignment of creating a CLI, because I think it would have helped me better understand how to implement rescuing exceptions- right now, I just created custom exceptions, but haven't yet implemented a course of action in how these errors are rescued.  |
| Describe a _nominal test_ that you wrote for this assignment. | One nominal test of my hotel class looked to see that the first reservation of a block was unassigned and then changed to assigned after the assign_block_reservation method.   |
| Describe an _edge case test_ that you wrote for this assignment. | I wrote an edge case test about the assign_block_reservation method that checked to see that, after three block reservations were change to "assigned", that the fourth tim the method was called, there was only 1 reservation left as "unassigned". This test was extremely helpful as I found out that my find_unassigned_block_reservations method in the code was not correctly identifying unassigned reservations. Before I fixed it, it would return all reservations in a block - the  assign_block_reservation method was not effectively changing the status of the reservations.   |
| How do you feel you did in writing pseudocode first, then writing the tests and then the code? | It's helpful to write out the intention and general idea of what you're trying to accomplish first; it's easy to just start coding away so writing pseudocode first helped to slow me down.   |

### CODE REVIEW

File: lib/reservation.rb [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+require 'awesome_print'
+require 'date'
+
+module Hotel_Chain
+  class Reservation
+
+    attr_accessor :check_in_date, :check_out_date, :room, :HOTEL, :cost, :status, :block_reserved
+
+    HOTEL = Hotel_Chain::MyHotel.new
```
Comment (CheezItMan) [inline lib/reservation.rb]: This looks like you can only have **ONE** hotel **EVER**.

File: lib/reservation.rb [hunk 2]
```diff
@@ -0,0 +1,32 @@
+#reservation.rb
+
+require 'awesome_print'
+require 'date'
+
+module Hotel_Chain
+  class Reservation
+
+    attr_accessor :check_in_date, :check_out_date, :room, :HOTEL, :cost, :status, :block_reserved
```
Comment (CheezItMan) [inline lib/reservation.rb:9]: You have a lot of `attr_accessor` methods here, but they don't all get set values, like `cost`.  How does that get calculated?

You should either have a method that does the calculation for the cost of the reservation or set the cost in the `initialize` method.

File: lib/hotel.rb [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+      return reservations_on_date
+    end
+
+    def print_reservations_by_date(date)
+      array = []
+      reservations_on_date = self.find_reservations_by_date(date)
+      reservations_on_date.each do |reservation|
+        array  << "Room #{reservation.room.room_id} is reserved from #{reservation.check_in_date} to #{reservation.check_out_date}"
+      end
+      ap array
```
Comment (CheezItMan) [inline lib/hotel.rb]: When you turn things in, it's best to take out extraneous `puts` or `ap` commands.  It makes running things more messy than it would be otherwise.

File: lib/hotel.rb [hunk 4]
```diff
@@ -0,0 +1,182 @@
+#Hotel.rb
+require 'awesome_print'
+require 'date'
+require 'pry'
+
+module Hotel_Chain
+  class MyHotel
```
Comment (CheezItMan) [inline lib/hotel.rb]: Why not call this `Hotel` instead of `MyHotel`?  `MyHotel` sounds like it's an individual `Hotel` instance rather than a class.

File: lib/hotel.rb [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+
+    attr_reader :array_of_rooms, :reservations_array, :blocks_array
```
Comment (CheezItMan) [inline lib/hotel.rb]: You should probably just name these, `rooms`, `reservations`, or `blocks` because you might in the future change how you implement these data structures and you want to separate implementation from how they're used.

Also do you want to give the user direct access to the rooms and reservations and blocks, instead of making them go through your methods.  With this they could change them directly bypassing `MyHotel` methods.

File: lib/hotel.rb [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+          list_array << "#{i+1}. Room #{object.room_id} - $#{object.rate}/night"
+          i += 1
+      end
+      return list_array
+    end
+
+    def find_reservations_by_date(date)
+      reservations_on_date = []
+      @reservations_array.each do |reservation|
+        if (reservation.check_in_date...reservation.check_out_date).cover?(Date.strptime(date, "%m/%d/%Y"))
```
Comment (CheezItMan) [inline lib/hotel.rb]: Just a note, this makes a new `Date` object each time the loop iterates.  You instead create one before the loop and then use it here.  

Just a bit of optimization.  

More seriously, this method works directly with the Reservation object's data attributes, instead this logic should go into the Reservation class and Hotel should simply ask, ""Hey does this date overlap with you?" using an instance method and should get back a true/false response.  

This is an example of tight coupling.

File: lib/hotel.rb [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+
+    def store_reservation(check_in_date, check_out_date)
+      available_rooms = []
+      #ap "STORE_RESERVATION START: reservations_array: #{reservations_array}"
+      if @reservations_array.length == 0 #i.e. there are no reservations at all
+        new_reservation = Hotel_Chain::Reservation.new(check_in_date, check_out_date)
+        new_reservation.room = @array_of_rooms[0]
+        @reservations_array << new_reservation
+        return new_reservation
+      elsif @reservations_array.length > 0
```
Comment (CheezItMan) [inline lib/hotel.rb]: This elseif is unnecessary, `else` would be enough since if it's not 0 it's greater than 0.

File: lib/hotel.rb [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+          new_reservation = Hotel_Chain::Reservation.new(check_in_date, check_out_date)
+          new_reservation.room = available_rooms[0]
+          @reservations_array << new_reservation
+          return new_reservation
+        end
+      end
+    end
+
+    def find_rooms_available(check_in_date, check_out_date)
+      unavailable_rooms = []
```
Comment (CheezItMan) [inline lib/hotel.rb:71]: This method seems overly complicated.

File: lib/hotel.rb [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+        end
+      end
+      final_available_rooms = @array_of_rooms - unavailable_rooms
+      return final_available_rooms
+    end
+
+
+    def reserve_block(party_name, check_in, check_out, no_of_rooms, room_rate)
+      if no_of_rooms > 5
+        raise ExceededRoomLimitForBlocksError
```
Comment (CheezItMan) [inline lib/hotel.rb:103]: Good use of a custom Error.

File: lib/hotel.rb [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+
+      local_reservation_array = []
+      available_rooms = find_rooms_available(check_in, check_out)
+
+      if available_rooms.length < no_of_rooms
+        raise ArgumentError.new("There are not enough rooms available to reserve that block")
+      else
+        no_of_rooms.times do |room|
+          new_reservation = store_reservation(check_in, check_out)
+          new_reservation.room.rate = room_rate
```
Comment (CheezItMan) [inline lib/hotel.rb:114]: room rate should be passed in when you create the reservation.

**Comment by CheezItMan** [comment]
# Hotel
## What We're Looking For

Feature | Feedback
---     | ---
**Baseline** |
Used git regularly	| A lot of commits and good messages, that's great! |
Answer comprehension questions	| Check, you did implement a rescue in your `Reservation` class which caught an ArgumentError and then generated a custom error of your own.   |
**Design** |
Each class is responsible for a single piece of the program | You've got almost all the functionality of your project in the `Hotel` class.  You should allocate actions that go along with the concept of reservation or block in those objects.  Like rooms remaining in a block, or total cost for a reservation.   |
Classes are loosely coupled | They're loosely coupled because none of the classes outside of Hotel is really doing business logic.   |
**Wave 1** |
List rooms | Check, but see my notes in the `Hotel` class. |
Reserve a room for a given date range | Check  |
List reservations for a given date | Check |
Calculate reservation price | Check, but it's done via very tight coupling between `Hotel` and `Reservation`. |
Invalid date range produces an error |  There's no check for if a reservation's check-in date is before the check-out date. |
**Wave 2** |
View available rooms for a given date range | Check  |
Reserving a room that is not available produces an error | It picks a random room, but it does check to see if none are available |
**Wave 3** |
Create a block of rooms | Check |
Check if a block has rooms | Check |
Reserve a room from a block | Check via `find_unassigned_block_reservations` |
Test coverage | 99%! |
**Additional Feedback** | Take a look at the Sandy Metz book again along with my in-line comments.  You've got some textbook examples of tight coupling, and I highlighted at least one.  You also have all the work being done in `MyHotel` instead of providing helper methods where appropriate in the other classes.  Think about what you should be able to ask a reservation or a block.  What questions should a reservation be able to answer?  Not a bad submission, but a difficult design to change or adapt due to all the work being very nested into `MyHotel`. |

**Review by CheezItMan — COMMENTED** [review]
A few in-code notes.
